### PR TITLE
Improve handling of waiting areas in FF Router

### LINKED ISSUE #796: Waiting areas depend on structure of the geometry
**JuPedSim - JPScore**

```
[2020-06-01 10:04:57.190] [info] Starting JuPedSim - JPScore
[2020-06-01 10:04:57.191] [info] Version 0.8.4
[2020-06-01 10:04:57.191] [info] Commit id v0.8.4-998-g61ce4d86-dirty
[2020-06-01 10:04:57.191] [info] Commit date Fri May 29 11:09:27 2020
[2020-06-01 10:04:57.191] [info] Build from branch 754-output_precision
[2020-06-01 10:04:57.191] [info] Build with clang++(11.0.0)
```

**Describe the bug**

I have the same geometry, but structed differently:

- Version 1: Three rooms. Every `room` has one `subroom`
- Version 2: One room with three subrooms.

**Expected behavior**
Simulation of two groups of pedestrians: 

- Agents in group 1: Go through the central room to outside.
- Agents in group 2: Go through the lower room to outside.

**Results of Version2 (nice)**

Screenshot 2020-06-01 10 12 36
Screenshot 2020-06-01 10 12 42

**Results of Version 1 (not nice)**

Screenshot 2020-06-01 10 09 53
Screenshot 2020-06-01 10 09 59

**Files:**
```
.
├── geometry1.xml
├── geometry2.xml
├── inifile.xml
└── inifile2.xml
```
[Issue.zip](https://github.com/JuPedSim/jpscore/files/4710162/Issue.zip)

### PR DESCRIPTION
**Pull request by schroedtert**
Before: Only the transition closest to a goal/waiting area was added to the list of potential candidates. Which lead to errors when the geometry only consists of one room with multiple subrooms.
Now: Depending on the exit strategy add all transition of the room containing the waiting area or all crossings and transition of the subroom containing the waiting area to the list of candidates. Also check if the pedestrian is in correct room/subroom before fast exit with the waiting area.

- solve bug described in #796 that the FF Router does not work correctly when splitting in the geometry into subrooms
- add tests to assure bug is fixed, test with room and subroom wise splitting of geometry and both FF Router exit strategies (8,9)
- some renaming to improve readablity

Closes #796

### CODE REVIEW

File: libcore/src/routing/ff_router/ffRouter.cpp [hunk 1]
```diff
@@ -120,56 +120,86 @@ void FFRouter::CalculateFloorFields()
...
+            if(auto * waitingArea = dynamic_cast<WaitingArea *>(goal); waitingArea != nullptr) {
+                auto * room = _building->GetRoom(waitingArea->GetRoomID());
+                std::set<int> doorUIDs;
+
+                if(!_targetWithinSubroom) {
+                    //candidates of current room (ID) (provided by Room)
+                    for(auto transUID : room->GetAllTransitionsIDs()) {
+                        doorUIDs.emplace(transUID);
+                    }
+                    for(auto const & [_, subroom] : room->GetAllSubRooms()) {
```
Comment (zeroset) [inline libcore/src/routing/ff_router/ffRouter.cpp:172]: I think we should decide for east const or west const and use it consistently.

File: libcore/src/routing/ff_router/ffRouter.cpp [hunk 2]
```diff
@@ -120,56 +120,86 @@ void FFRouter::CalculateFloorFields()
...
+                        for(auto const * cross : subroom->GetAllCrossings()) {
+                            doorUIDs.emplace(cross->GetUniqueID());
+                        }
+                    }
+                } else {
+                    //candidates of current subroom only
+                    auto subroom = room->GetSubRoom(waitingArea->GetSubRoomID());
+                    for(const auto & crossing : subroom->GetAllCrossings()) {
```
Comment (zeroset) [inline libcore/src/routing/ff_router/ffRouter.cpp:180]: decide where to put the const ;).

File: libcore/src/routing/ff_router/ffRouter.cpp [hunk 3]
```diff
@@ -120,56 +120,86 @@ void FFRouter::CalculateFloorFields()
...
+                auto * room = _building->GetRoom(waitingArea->GetRoomID());
+                std::set<int> doorUIDs;
+
+                if(!_targetWithinSubroom) {
+                    //candidates of current room (ID) (provided by Room)
+                    for(auto transUID : room->GetAllTransitionsIDs()) {
+                        doorUIDs.emplace(transUID);
+                    }
+                    for(auto const & [_, subroom] : room->GetAllSubRooms()) {
+                        for(auto const * cross : subroom->GetAllCrossings()) {
```
Comment (zeroset) [inline libcore/src/routing/ff_router/ffRouter.cpp:173]: decide where to put the const ;).

File: libcore/src/routing/ff_router/ffRouter.cpp [hunk 4]
```diff
@@ -120,56 +120,86 @@ void FFRouter::CalculateFloorFields()
...
+                            doorUIDs.emplace(cross->GetUniqueID());
+                        }
+                    }
+                } else {
+                    //candidates of current subroom only
+                    auto subroom = room->GetSubRoom(waitingArea->GetSubRoomID());
+                    for(const auto & crossing : subroom->GetAllCrossings()) {
+                        doorUIDs.emplace(crossing->GetUniqueID());
+                    }
+                    for(const auto & transition : subroom->GetAllTransitions()) {
```
Comment (zeroset) [inline libcore/src/routing/ff_router/ffRouter.cpp:183]: decide where to put the const ;).

File: libcore/src/routing/ff_router/ffRouter.cpp [hunk 5]
```diff
@@ -120,56 +120,86 @@ void FFRouter::CalculateFloorFields()
...
+                    }
                 }
+
+                _doorsToGoalUID.emplace(goalID, doorUIDs);
+            } else {
+                // if goal find only closest exit
+                double minDist = std::numeric_limits<double>::max();
+                int minID      = -1;
+
+                for(auto const & [exitID, exit] : _exitsByUID) {
```
Comment (zeroset) [inline libcore/src/routing/ff_router/ffRouter.cpp:194]: decide where to put the const ;).

File: libcore/src/routing/ff_router/ffRouter.cpp (hunk 1 above)
Comment (schroedtert) [reply libcore/src/routing/ff_router/ffRouter.cpp]: Changed, to be consistent

File: libcore/src/routing/ff_router/ffRouter.cpp (hunk 2 above)
Comment (schroedtert) [reply libcore/src/routing/ff_router/ffRouter.cpp:180]: Not changed

File: libcore/src/routing/ff_router/ffRouter.cpp (hunk 3 above)
Comment (schroedtert) [reply libcore/src/routing/ff_router/ffRouter.cpp]: Changed, to be consistent

File: libcore/src/routing/ff_router/ffRouter.cpp (hunk 4 above)
Comment (schroedtert) [reply libcore/src/routing/ff_router/ffRouter.cpp:183]: Not changed

File: libcore/src/routing/ff_router/ffRouter.cpp (hunk 5 above)
Comment (schroedtert) [reply libcore/src/routing/ff_router/ffRouter.cpp]: Changed, to be consistent

**Pull request merged**